proposals/0177-program-runtime-abiv2.md

```diff
@@ -0,0 +1,289 @@
+  - Reference to a slice of instruction accounts `&[InstructionAccount]`, 
+    consisting of:
+    - Pointer to beginning of slice: `u64` (points to the `0x700000000` 
+      region)
+    - Number of elements in slice: `u64`
+  - Instruction data `&[u8]`, which is composed of:
+    - Pointer to data: `u64`
+    - Length of data: `u64`
+
+Let `InstrucionAccount` contain the following fields:
+
+  - Index to transaction account: `u16`
+  - Signer flag: `u8` (1 for signer, 0 for non-singer)
+  - Writable flag: `u8` (1 for writable, 0 for readonly)
+
+The second region is readonly and starts at `0x700000000`. It must also be 
+updated at each CPI call edge. This region contains an array of all 
+`InstructionAccounts` for every instruction in the transction. In other words:
+
+- For instruction in transaction:
+  - For each account in instruction:
+    - `InstructionAccount`, consisting of:
+      - Index to transaction account: `u16`
+        - Signer flag: `u8` (1 for signer, 0 for non-singer)
+        - Writable flag: `u8` (1 for writable, 0 for readonly)
+
+#### Return data scratchpad
+
+A writable memory region starting at `0x800000000` must be mapped in for the
+return-data scratchpad.
+
+### Accounts area
+
+For each unique (meaning deduplicated) instruction account the payload must
+be mapped in at `0x900000000` plus `0x100000000` times the index of the
+**transaction** account (not the index of the instruction account). Only if the
+instruction account has the writable flag set and is owned by the current
+program it is mapped in as a writable region. The writability of a region must
+be updated as programs through out the transaction modify the account metadata.
+
+The runtime must only map the payload for accounts that belong in the current
+executing instruction. The payload for accounts belonging to sibling instructions
+must NOT be mapped.
+
+### Instruction payload area
+
+For each instruction, the runtime must map its payload at address 
+`0x10900000000` plus `0x100000000` times the index of the instruction in the 
+trasaction. All instruction payload mappings are readonly.
+
+One extra writable mapping must be created after the last instruction payload 
+area to be the CPI scratch pad, i.e. at address `0x10900000000` plus 
+`0x100000000` times the number of instructions in the transaction. Its purpose 
+is for programs to write CPI instruction data directly to it and avoid copies.
+
+### VM initialization
+
+During the initilization of the virtual machine, the runtime must load the 
+value `0x400000000` in register `R1`, value `0x500000000` in register `R2`, 
+and value `0x600000000` in register `R3`. These values represent addresses for 
+programs to easily find the areas to read information about the transaction 
+and the instructions.
+
+### Changes to syscalls
+
+Changes to the account metadata must now be communicated with specific 
+syscalls, as detailed below:
+
+- `sol_assign_owner`: Dst account, new owner as `&[u8; 32]`
```

Comment:
This sys-call makes the owner field volitile and prevents references to it from a program writing perspective. Still working on a way around this, but may require changing the signature to take in a pointer to the owner field to make it non-volitile (cpi still has issues here though).

Reply:
> may require changing the signature to take in a pointer to the owner field to make it non-volitile (cpi still has issues here though).

That is a relevant point. I think we can change it to be a pointer yes.

Regarding `sol_transfer_lamports`, would making the parameter a pointer also help? Or could we get around using volatile there?

Reply:
I was thinking through this and I think the only solution due to the volatile memory is to make a comparison syscall as @febo described. Otherwise there will always have to be a copy of the owner field. The other method would be to change the api to be non-voltile on the owner field and owner updates must be obtained from another syscall.